### PR TITLE
Added shrink to Vector and RVector

### PR DESCRIPTION
**Pull request by wintersteiger**

### CODE REVIEW

File: ulib/LowStar.Vector.fst [hunk 1]
```diff
@@ -497,6 +498,26 @@ let flush #a vec ia i =
...
+  HST.ST (vector a)
+    (requires (fun h0 ->
+      live h0 vec /\ freeable vec /\
+      HST.is_eternal_region (frameOf vec)))
+    (ensures (fun h0 fvec h1 ->
+      frameOf vec = frameOf fvec /\
+      hmap_dom_eq h0 h1 /\
+      live h1 fvec /\ freeable fvec /\
+      modifies (loc_union (loc_addr_of_vector vec)
+                          (loc_vector fvec)) h0 h1 /\
```
Comment (nikswamy) [inline ulib/LowStar.Vector.fst]: Why the modifies clause? This code is actually pure, right?

Comment (wintersteiger) [reply ulib/LowStar.Vector.fst]: Good point! It's just there because I copied it from `flush`; removed now.

**Comment by joonwonc** [comment]
In my opinion, it's more natural for `LowStar.Vector.shrink` to behave similar to what `LowStar.Vector.flush` does -- to free the original store and to reallocate a smaller space.
`LowStar.RVector.shrink` (that employes `LowStar.Vector.shrink`) frees each deprecated *element* while shrinking, so it seems a bit unnatural to hold such invalid pointers in `vector`, even if they cannot be accessed by `vector` APIs (because we put smaller `Vec?.sz` those pointers cannot be accessed).

`hmap_dom_eq` is to ensure that a `vector` operation does not add (or remove) new regions, which is required in `LowStar.RVector`.

**Comment by wintersteiger** [comment]
`shrink` is just supposed to make the vector smaller. The elements are in `vs` and we don't have to free them immediately, they can keep their allocated memory, so `cap` stays the same, while `sz` becomes smaller. So far I'm not seeing any memory leaks, so it seems to work.

Should `flush` and `shrink` perhaps return `unit` and modify the existing vector?

**Comment by msprotz** [comment]
please do not rebase this pull request I have merged it on my local branch

**Comment by nikswamy** [comment]
I don't think `flush` and `shrink` can return unit and mutate the existing vector; not without revising the library as discussed in the original pull request for vectors to add a level of indirection in the representation of vectors.

Can someone explain why it is desirable to consume the original vector, preventing further use of it, even though shrink doesn't actually free anything?

**Comment by aseemr** [comment]
It creates an aliasing of the underlying buffer. Both the original vector and the new vector -- in the current implementation -- share the buffer, while the spec doesn't say that they are aliased. So consuming liveness of the input buffer seems necessary.

However, we could reflect this aliasing by saying `loc_vector vec == loc_vector new_vec /\ loc_addr_of_vector vec == loc_addr_of_vector new_vec`, and then say `h0 == h1`. That way `hmap_dom_eq` would also be unnecessary.

If `h0 == h1` is not what you want, `hmap_dom_eq` is still lower abstraction than what the clients should be using. The code should be able to derive that a region is not freed by using the `modifies` clause (if not, we should fix it). Is the code also sensitive to additions of regions?

**Comment by wintersteiger** [comment]
@nikswamy fair enough, the library should definitely be kept consistent. My reason for keeping the old vector is performance - shrinking will happen frequently (in my application) and I'd rather reuse the elements instead of copying them around, just to see the old vector go out of scope right after.

@aseemr I like your `h0 == h1` idea, that sounds like it could simplify things downstream. If, like in my case, the old vectors aren't used anymore, having that aliasing constraint around is no complication at all.

**Comment by aseemr** [comment]
The library is quite transparent, how about writing `shrink` as just a `Tot` function instead:

```
val shrink:
  #a:Type -> vec:vector a ->
  new_size:uint32_t{new_size <= size_of vec} ->
  Tot (vector a)
let shrink #a vec new_size =
  Vec new_size (Vec?.cap vec) (Vec?.vs vec)
```

**Comment by nikswamy** [comment]
Sorry @aseemr: I still don't see why consuming the liveness of the input vector is necessary. I'm actually very confused about the whole point of the `LowStar.Vector.shrink` operation. Let me try ... 

* `shrink` is morally a pure operation: it just returns a vector that shares storage with the input vector, but the returned vector has a smaller current size than the input one.

* We must specify that the returned vector is aliased with the input vector: otherwise, from a client's perspective, any subsequent mutation of the contents of the returned vector can interfere with the contents of any other location in the same region.

* If we must specify the the returned vector aliases the original one, then the original one must be live too: we cannot consume it.

So, I see one of Aseem's proposal as compatible with all of this:

> However, we could reflect this aliasing by saying `loc_vector vec == loc_vector new_vec /\ loc_addr_of_vector vec == loc_addr_of_vector new_vec`, and then say `h0 == h1`.

But then, what's this `shrink` operation really achieving?

If anything, one might hope to use it for some kind of capability based access control. For instance,

1. I have a vector `v`
2. I shrink it to `u`, which is a prefix of `v`
3. I hand `u` off to some piece of code and that code claims to only modify `loc_vector u`
4. I'd like to conclude that the suffix of `v` is unmodified (and ideally, not even read, but that's a separate discussion)

But, even this doesn't quite work, because `loc_vector v == loc_vector u` (and must inherently be so, since they share the underlying storage and `u` can be expanded once again to write beyond its current size)

So, what's `shrink` good for?

**Comment by aseemr** [comment]
Use cases for `shrink` is a question for @wintersteiger . But if we keep it, I think we can have it as a `Tot` function. All the preconditions seem unnecessary.

**Comment by wintersteiger** [comment]
We need this for the Merkle trees (in Coco). When a transaction/commit fails (and similar), we have to backtrack to an older state and up to now we kept separate copies of old trees, which is way too expensive, while it's technically trivial and cheap to just truncate/shrink the vectors used to represent the trees.

**Comment by wintersteiger** [comment]
Guys, we need this fairly soon. I kept the modifies clause for now because the rest of the proofs work with it. Parts of the Merkle trees are now fragmented between hacl-star branches and we need to merge all of this. I'd appreciate if you could take another look at it. Feel free to commit directly to the PR branch or to close the PR and put it elsewhere!

**Comment by nikswamy** [comment]
I still don't fully understand what the truncation is buying you. But, that's ok.

Aseem's suggestion of using a `Tot` function for this makes good sense to me. The current spec in Stack has too many rough edges

**Comment by aseemr** [comment]
We discussed it briefly in last F* call and Nik also seemed ok with making it a Tot function. Could you please try that. It would be much simpler also.

**Comment by wintersteiger** [comment]
@nikswamy truncation is buying me that I don't have to copy a vector. Is there another way to do that?

**Comment by wintersteiger** [comment]
@aseemr the Tot version works great! I have no idea where it gets all the invariant proofs from, but it does get them from somewhere! I suppose that means that I can simplify `RVector.shrink` as well?

**Comment by joonwonc** [comment]
`RVector.shrink` `free`s some elements so I think it cannot be a `Tot` function..?

**Comment by aseemr** [comment]
It can't be a `Tot` functions, but the proof can be simplified, e.g. `hh1` and `hh2` are now syntactically the same.

The spec of `RVector.shrink` has the same issues though: we don't know anything about the `loc` of the returned Vector. I think we should also add aliasing postconditions between `rv` and `frv` to the `ensures` spec.

**Comment by nikswamy** [comment]
@aseemr I had a look and it seems like the aliasing spec is `RVector.shrink` is fine. The intended use of this library is to use regions as the granularity of aliasing. e.g., `loc_rvector` is defined as the entire region that holds the vector. In this case, `RVector.shrink` says that the returned vector is in the same region as the original one and as such aliases it.

I'm going to kick another CI run and try to merge it.

**Pull request merged**